File: root/lang/javascript/__demo__/algo-demo/src/list/singly_linked_list.ts
import { type IList } from './index'

class ListNode<T> {
  data: T
  next?: ListNode<T>

  constructor(data: T) {
    this.data = data
  }
}

export class LinkedList<T> implements IList<T> {
  private head: ListNode<T> | undefined
  size: number

  constructor() {
    this.size = 0
  }

  pushFront(data: T): void {
    const newNode = new ListNode<T>(data)
    newNode.next = this.head
    this.head = newNode
  }

  pushBack(data: T): void {
    const newNode = new ListNode<T>(data)
    if (this.head === undefined) {
      this.head = newNode
    } else {
      let curr = this.head
      while (curr.next !== undefined) {
        curr = curr.next
      }
      curr.next = newNode
    }
    this.size++
  }

  popFront(): T | undefined {
    if (this.head === undefined) {
      return undefined
    } else {
      const front = this.head.data
      this.head = this.head.next
      this.size--
      return front
    }
  }

  popBack(): T | undefined {
    if (this.head === undefined) {
      return undefined
    } else {
      let curr = this.head
      let prev = undefined
      while (curr.next !== undefined) {
        prev = curr
        curr = curr.next
      }
      if (prev === undefined) {
        this.head = undefined
      } else {
        prev.next = undefined
      }
      this.size--
      return curr.data
    }
  }

  peekFront(): T | undefined {
    return this.head?.data
  }

  peekBack(): T | undefined {
    if (this.head === undefined) {
      return undefined
    } else {
      let curr = this.head
      while (curr.next !== undefined) {
        curr = curr.next
      }
      return curr.data
    }
  }

  get(index: number): T {
    this.checkIndex(index)
    let curr = this.head
    for (let i = 0; i < index; i++) {
      curr = curr?.next
    }
    curr = curr as ListNode<T>
    return curr.data
  }

  set(index: number, data: T): void {
    this.checkIndex(index)
    let curr = this.head
    for (let i = 0; i < index; i++) {
      curr = curr?.next
    }
    curr = curr as ListNode<T>
    curr.data = data
  }

  del(index: number): void {
    this.checkIndex(index)
    let curr = this.head
    let prev = undefined
    let next = curr?.next
    for (let i = 0; i < index; i++) {
      prev = curr
      curr = curr?.next
      next = curr?.next
    }
    curr = curr as ListNode<T>
    if (prev === undefined) {
      this.head = next
    } else {
      prev.next = next
    }
    this.size--
  }

  add(index: number, data: T): void {
    this.checkIndex(index)
    const newNode = new ListNode<T>(data)
    let curr = this.head
    let prev = undefined
    for (let i = 0; i < index; i++) {
      prev = curr
      curr = curr?.next
    }
    curr = curr as ListNode<T>
    newNode.next = curr
    if (prev === undefined) {
      this.head = newNode
    } else {
      prev.next = newNode
    }
    this.size++
  }

  reverse(): void {
    let curr = this.head
    let prev = undefined
    while (curr !== undefined) {
      const next = curr.next
      curr.next = prev
      prev = curr
      curr = next
    }
    this.head = prev
  }

  toString(): string {
    const result = []
    let curr = this.head
    while (curr !== undefined) {
      result.push(curr.data)
      curr = curr.next
    }
    return JSON.stringify(result)
  }

  private checkIndex(index: number): void {
    if (index < 0 || index >= this.size) {
      throw new RangeError('Index out of bounds')
    }
  }
}

File: root/lang/rust/__demo__/algo-demo/src/list/singly_linked_list.rs
struct Node<T> {
    data: T,
    next: Option<Box<Node<T>>>,
}

pub struct LinkedList<T> {
    head: Option<Box<Node<T>>>,
}

pub struct IntoIter<T>(LinkedList<T>);

impl<T> Node<T> {
    fn new(data: T) -> Self {
        Self { data, next: None }
    }
}

impl<T> LinkedList<T> {
    pub fn new() -> Self {
        Self { head: None }
    }

    pub fn push_front(&mut self, data: T) {
        let mut new_node = Box::new(Node::new(data));
        new_node.next = self.head.take();
        self.head = Some(new_node);
    }

    pub fn pop_front(&mut self) -> Option<T> {
        self.head.take().map(|it| {
            self.head = it.next;
            it.data
        })
    }

    pub fn peek_front(&self) -> Option<&T> {
        self.head.as_ref().map(|it| &it.data)
    }

    pub fn peek_front_mut(&mut self) -> Option<&mut T> {
        self.head.as_mut().map(|it| &mut it.data)
    }

    pub fn push_back(&mut self, data: T) {
        let new_node = Box::new(Node::new(data));
        match self.head.as_mut() {
            None => {
                self.head = Some(new_node);
            }
            Some(mut current) => {
                while let Some(ref mut next) = current.next {
                    current = next;
                }
                current.next = Some(new_node);
            }
        }
    }

    pub fn pop_back(&mut self) -> Option<T> {
        None
    }

    pub fn reverse(&mut self) {
        let mut prev = None;
        while let Some(mut current) = self.head.take() {
            self.head = current.next.take();
            current.next = prev;
            prev = Some(current);
        }
        self.head = prev;
    }
}

impl<T: std::fmt::Display> std::fmt::Display for LinkedList<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut node = &self.head;
        while let Some(current) = node {
            write!(f, "{} -> ", current.data)?;
            node = &current.next;
        }
        write!(f, "None")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_push_front() {
        let mut list = LinkedList::new();
        (1..=5).for_each(|x| list.push_front(x));
        println!("{list}");
    }

    #[test]
    fn test_push_back() {
        let mut list = LinkedList::new();
        (1..=5).for_each(|x| list.push_back(x));
        println!("{list}");
    }

    #[test]
    fn test_reverse() {
        let mut list = LinkedList::new();
        (1..=5).for_each(|x| list.push_back(x));
        list.reverse();
        println!("{list}");
    }
}
